cpp/src/interop/to_arrow_schema.cpp

```diff
@@ -171,7 +171,7 @@ int dispatch_to_arrow_type::operator()<cudf::list_view>(column_view input,
   auto child = input.child(cudf::lists_column_view::child_column_index);
   ArrowSchemaInit(out->children[0]);
   auto child_meta =
-    metadata.children_meta.empty() ? column_metadata{"element"} : metadata.children_meta[0];
+    metadata.children_meta.empty() ? column_metadata{"element"} : metadata.children_meta[1];
```

Comment:
For my education, what is this change for?

Reply:
I have the same question here, shouldn't the metadata for the list's child be at index 0, not 1?

Reply:
As I think about this a bit more, I am leaning towards the answer being yes. The reason for this change is that by convention the children metadata in a `column_metadata` object correspond 1-1 with the children of a nested column. For list columns, the 0th child is the offsets, so we have historically always included an empty `column_metadata` instance corresponding to that. That convention is really only established by our previous `to_arrow` implementation, though, and I don't see any reason that we can't change it.

The argument in favor of changing would be that the current behavior violates the principle of least surprise. Users probably aren't expecting to have to stick in an empty entry to account for the offsets in a list column. The argument against changing is that it would break the symmetry of the API to say that there's a 1-1 correspondence with the number of child columns, _except_ for list columns where we ignore one of the children altogether. I'm leaning towards the former being the better outcome, though.

What do you two (and @bdice) think?

Reply:
We discussed this a bit offline since I lacked the context to understand what was happening. It seems like 0 was a bug, and this should've been 1. However, we contemplated the possibility of not constructing a metadata object for the offsets column because it's not useful. That would then make the correct index 0 again (but it would correspond to the child data metadata instead of the offsets if we don't create offset metadata).

I vote in favor of keeping this as 1 but writing it as `cudf::lists_column_index::child_column_index` instead of hardcoding `1`. If we want to skip construction of the offsets metadata we can do that in a later PR. For now, I think it is clearer to libcudf developers that `cudf::lists_column_index::child_column_index` has more meaning.

Reply:
Always use these `static constexpr` defined in these column types. It helps us during refactoring too.
eg. `cudf::lists_column_index::child_column_index`, `cudf::dictionary_column_view::indices_column_index`, `cudf::strings_column_view::offsets_column_index`

Declare any magic constants as `static constexpr` inside a class, or `constexpr` in header/source files.

Reply:
I'm generally in favor of the principle of least surprise. I'd definitely agree with not requiring accounting for the offsets with an empty metadata value. That's my opinion anyways, but this makes sense now.

Reply:
Opened https://github.com/rapidsai/cudf/issues/16600 to keep this in mind as a future improvement after the dust settles a little here.